tdrs-backend/tdpservice/parsers/parse.py

```diff
@@ -0,0 +1,131 @@
+"""Convert raw uploaded Datafile into a parsed model, and accumulate/return any errors."""
+
+
+import os
+from . import schema_defs
+from . import validators
+
+
+def parse_datafile(datafile):
+    """Parse and validate Datafile header/trailer, then select appropriate schema and parse/validate all lines."""
+    rawfile = datafile.file
+    errors = {}
+
+    document_is_valid, document_error = validators.validate_document(rawfile)
+    if not document_is_valid:
+        errors['document'] = [document_error]
+        return errors
+
+    # get header line
+    rawfile.seek(0)
+    header_line = rawfile.readline().decode().strip()
+
+    # get trailer line
+    rawfile.seek(0)
+    rawfile.seek(-2, os.SEEK_END)
+    while rawfile.read(1) != b'\n':
+        rawfile.seek(-2, os.SEEK_CUR)
+
+    trailer_line = rawfile.readline().decode().strip('\n')
+
+    # parse header, trailer
+    header, header_is_valid, header_errors = schema_defs.header.parse_and_validate(header_line)
+    if not header_is_valid:
+        errors['header'] = header_errors
+
+    trailer, trailer_is_valid, trailer_errors = schema_defs.trailer.parse_and_validate(trailer_line)
+    if not trailer_is_valid:
+        errors['trailer'] = trailer_errors
+
+    if not header_is_valid or not trailer_is_valid:
+        return errors
+
+    # determine the file type
+    schema_options = get_schema_options(header['program_type'])
+
+    # ensure file section matches upload section
+    section_names = {
+        'A': 'Active Case Data',
+        'C': 'Closed Case Data',
+        'G': 'Aggregate Data',
+        'S': 'Stratum Data',
+    }
+
+    section = header['type']
+
+    if datafile.section.find(section_names[section]) == -1:
+        errors['document'] = ['Section does not match.']
+        return errors
+
+    # parse line with appropriate schema
+    rawfile.seek(0)
+    line_number = 0
+
+    for rawline in rawfile:
+        line_number += 1
+        line = rawline.decode().strip('\r\n')
+
+        if line.startswith('HEADER') or line.startswith('TRAILER'):
+            continue
```

Comment:
Why do we skip these lines? Is it because header/trailer are universal so would over-complicate the `get_schema` call?

Reply:
the header and trailer lines are parsed separately (lines 32-37), namely so that the `type` and `program_type` can be extracted from the header and used to select the schema to use. since the trailer isn't really used in preparsing, we could skip it in the preparsing validation and allow it to be parsed as part of the main loop. however, i think the header still needs to be preparsed to select the schema

Reply:
I believe we covered this on Friday so I'm fine leaving it as is but did want to note that we aren't backed into a corner as the program_type is only available from the header line -- all header data points are stored in the datafile model's members.